packages/react-server/src/ReactFlightServer.js

```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
   +[key: string]: ReactModel,
 };
 
+type Segment = {
+  id: number,
+  model: ReactModel,
+  ping: () => void,
+};
+
 type OpaqueRequest = {
   destination: Destination,
-  model: ReactModel,
-  completedChunks: Array<Uint8Array>,
+  nextChunkId: number,
+  pendingChunks: number,
+  pingedSegments: Array<Segment>,
+  completedJSONChunks: Array<Uint8Array>,
+  completedErrorChunks: Array<Uint8Array>,
   flowing: boolean,
+  toJSON: (key: string, value: ReactModel) => ReactJSONValue,
 };
 
 export function createRequest(
   model: ReactModel,
   destination: Destination,
 ): OpaqueRequest {
-  return {destination, model, completedChunks: [], flowing: false};
+  let pingedSegments = [];
+  let request = {
+    destination,
+    nextChunkId: 0,
+    pendingChunks: 0,
+    pingedSegments: pingedSegments,
+    completedJSONChunks: [],
+    completedErrorChunks: [],
+    flowing: false,
+    toJSON: (key: string, value: ReactModel) =>
+      resolveModelToJSON(request, value),
+  };
+  request.pendingChunks++;
+  let rootSegment = createSegment(request, model);
+  pingedSegments.push(rootSegment);
+  return request;
+}
+
+function attemptResolveModelComponent(element: React$Element<any>): ReactModel {
+  let type = element.type;
+  let props = element.props;
+  if (typeof type === 'function') {
+    // This is a nested view model.
+    return type(props);
+  } else if (typeof type === 'string') {
+    // This is a host element. E.g. HTML.
+    return renderHostChildrenToString(element);
+  } else {
+    throw new Error('Unsupported type.');
+  }
+}
+
+function pingSegment(request: OpaqueRequest, segment: Segment): void {
+  let pingedSegments = request.pingedSegments;
+  pingedSegments.push(segment);
```

Comment:
Could this segment already be in the array? What would happen in that case? Is it okay to have it twice?

Reply:
It’s ok to have it twice but it shouldn’t happen since a component only throws one promise at a time.